server/src/main/java/org/elasticsearch/Build.java

```diff
@@ -246,19 +281,30 @@ public static void writeBuild(Build build, StreamOutput out) throws IOException
      * @return the fully qualified build
      */
     public String qualifiedVersion() {
-        return version;
+        return qualifiedVersionString(version, qualifier, isSnapshot);
+    }
+
+    private static String qualifiedVersionString(String version, String qualifier, boolean isSnapshot) {
+        String versionString = version;
+        if (qualifier != null) {
```

Comment:
I think we should maintain one set of semantics for this, and I would rather not have that be two different ways of defining an empty qualifier. I went with `null` as the non-existent qualifier because it matches nicely with both getProperty and Matcher.group. As far as empty string, it's not possible because the build will fail if the qualifier doesn't match the qualifier pattern (see VersionPropertiesLoader where it is read and validated). Similarly, when reading a Build from an older node, the regex does not allow an empty qualifier (either doesn't exist, resulting in null, or it matches the pattern and is non-empty).

Reply:
This is what I see in Matcher.group javadocs:

> If the match was successful but the group specified failed to match any part of the input sequence, then null is returned

There is also a note about returning empty string, but that is only if the pattern itself allows the empty string, which this does not.